### PR TITLE
Vibcreg

### PR DESCRIPTION
**Pull request by Froskekongen**
Implementation of VIbCReg (https://arxiv.org/abs/2109.00783), which was originally developed for times series modeling.

The performance without tuning parameters and using the same parameters as VICReg gets a performance of 91.18 acc@1 online, and 99.74 acc@5 online.

### CODE REVIEW

**Comment by vturrisi** [comment]
Hey @Froskekongen thanks a lot for contributing. I'll take a look at the paper first to get a general idea.
It might take longer than normal because of CVPR, but I'll try to give some feedback as soon as I can.

**Comment by DonkeyShot21** [comment]
Hi @Froskekongen! We appreciate your initiative. 

Did you happen to run CIFAR100 as well?

**Comment by Froskekongen** [comment]
@DonkeyShot21: I will run CIFAR-100 overnight.

**Review by vturrisi — CHANGES_REQUESTED** [review]
I found some small things that would be good to be fixed before merging.
Also, can you add the loss to `solo/losses/__init__.py`?

TODO:
- implement tests
- add method to the docs
- add tests
- add results/checkpoints to README
- black files (the bot seems to have failed for some reason)

File: solo/utils/whitening.py [hunk 1]
```diff
@@ -66,3 +68,167 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+                    batch1=torch.matrix_power(P[k], 3),
+                    batch2=Sigma_N,
+                )
+            saved.extend(P)
+            wm = P[ctx.T].mul_(
+                rTr.sqrt()
+            )  # whiten matrix: the matrix inverse of Sigma, i.e., Sigma^{-1/2}
+
+            running_mean.copy_(momentum * mean + (1.0 - momentum) * running_mean)
+            running_wmat.copy_(momentum * wm + (1.0 - momentum) * running_wmat)
```
Comment (vturrisi) [inline solo/utils/whitening.py:113]: I'm getting this error. It's probably related to the `clone()` that I said we could remove. Sorry about that.

`E           RuntimeError: unsupported operation: more than one element of the written-to tensor refers to a single memory location. Please clone() the tensor before performing the operation.`

**Comment by vturrisi** [comment]
@Froskekongen I cleaned a bit, added some placeholders for results in the readme and fixed part of the tests. We are still getting an error which I assume is due to the `clone()` that I asked you to remove. We are also missing documentations (me or @DonkeyShot21 can add later) and the results/checkpoints for CIFAR10/100 and Imagenet-100.

**Comment by vturrisi** [comment]
I'll take care of the documents. @Froskekongen are you running/planning to run the experiments?

**Comment by Froskekongen** [comment]
@vturrisi: I ran the experiments for cifar10 and cifar100
cifar10: val_acc1 - 91.18, val_acc5 - 99.74
cifar100: val_acc1 - 67.37, val_acc5 - 90.07

**Comment by vturrisi** [comment]
Ok! I'll create some folders so that you can upload the checkpoints, k? About imagenet100, are you running or planning to run?

**Comment by Froskekongen** [comment]
Yes, I can upload the trained models to some folders. I didn't plan to run imagenet100 - at least not right now.

File: solo/utils/whitening.py (hunk 1 above)
Comment (Froskekongen) [reply solo/utils/whitening.py:113]: I reintroduced the .clone().

**Comment by vturrisi** [comment]
@Froskekongen just to confirm we are missing the checkpoints and Imagenet-100 results.
You can add the cifar10 checkpoint here: https://drive.google.com/drive/folders/19u3p1maX3xqwoCHNrqSDb98J5fRvd_6v?usp=sharing and cifar100 here: https://drive.google.com/drive/folders/19u3p1maX3xqwoCHNrqSDb98J5fRvd_6v?usp=sharing, both the `args.json` file and `the checkpoint itself`. When I'm able to run imagenet-100, I'll upload everything and we merge the PR.

**Comment by Froskekongen** [comment]
I have added the args and checkpoints to the drive folder.

**Comment by vturrisi** [comment]
Hey @Froskekongen. I'll run on imagenet-100. Can I use the same parameters as in VICReg?

**Comment by Froskekongen** [comment]
@vturrisi: The only thing that needs to be changed are the loss weights, which should be:
```
    --sim_loss_weight 25.0 \
    --var_loss_weight 25.0 \
    --cov_loss_weight 200.0 \
```

**Comment by vturrisi** [comment]
@Froskekongen just started with imagenet-100. When this is done and I run offline linear eval, I'll merge the PR (that should be tomorrow or Monday).

**Review by vturrisi — APPROVED** [review]

**Comment by vturrisi** [comment]
Merging it now. Thanks again @Froskekongen for contributing.

**Pull request merged**